### ISSUE TITLE
Peacock without a workspace: Allow option for colors to only affect one user and one instance

### ISSUE THREAD

**Issue by johnpapa** [johnpapa/vscode-peacock#7]
Currently the extension sets the colors in workspace settings. THis means whatever settings are in that file are then either gitignored or allowed, and then team members will also get your color. It may be desirable for many to only let the color affect one instance of vscode and for a specific user.

**Comment by burkeholland** [comment]
This is kinda related to #8.

I think it's ok to set it in the workspace settings. If you set any other colors there, they will also change for other people.

**Comment by johnpapa** [comment]
But is that undesirable? Imagine I change to bright yellow and you get it too, you may not like that?

**Comment by burkeholland** [comment]
This is true, but if you change any other color in the editor, I will get that too. What makes the title bar special?

**Comment by johnpapa** [comment]
Good question. Let me clarify a little.

If you change a color and put it in user settings, it is your color setting and applies only to you, but across all of your VS Code instances.

If you change a color and put it in workspace settings, it is (by default) a changed file that will go to github and apply to anyone who gets your repo, only for VS Code instances opening that code base.


The workspace settings will affect others, unknowingly.

The user settings don't do that ^^ , but won't let you color each code instance


Does that make sense?

**Comment by burkeholland** [comment]
What if we allow both? Set it for everyone in user settings, or read from editorconfig? This would address #8 as well and is something I can work on.

**Comment by johnpapa** [comment]
Interesting. But it won’t solve the original problem of allowing you to color one instance and not impose it on others. If it goes in user settings then every code window for you would be the same color still

**Comment by crossan007** [comment]
What about adding a key-value pair list to the user settings value; indicating a color for each "known" workspace file (maybe by workspace name)?

**Comment by johnpapa** [comment]
Had this neat idea from the Code team

```
You could easily set a user setting like:

  "peacock" : {
    "/development/vscode-browserpreview" : {
      "titleBar.activeBackground": ""
    }
  }

And namespace/scope the settings to a particular workspace path. Not shared and local to the user.

------
Kenneth Auchenberg
```

**Comment by sdsalyer** [comment]
Hi there. :) Great idea for an extension @johnpapa . Kind of in the vein of this thread:

1.  is it possible to just set a temporary color (ie. per session / in memory) without creating the .vscode/settings.json in the folder directory? 
2. It might also be useful to have per-language settings (e.g. SQL files are red, c# files are blue, etc.). 

Thanks for listening!

**Comment by johnpapa** [comment]
Thanks.

1.  hmmm. i dont think so. i think it needs to be in the settings.

2. The idea for coloring files (maybe the tabs) has come up (somewhere in here or twitter). I'm not sure it would be fun nor useful to look at a code instance with files tabs of red, yellow, blue, etc ... maybe too much color then. or did i miss your intention?

**Comment by sdsalyer** [comment]
re: 2, I guess I was thinking the active file's language association would
color the title bar, etc. (not necessarily the tabs -- I agree that would
be unpleasant with more than a couple colors). I suppose it could get
rather hectic, though, if switching between different languages regularly.
It was just a thought.

If you figure out #1 though, temporarily changing the color setting, that
would be boss. :-)

**Comment by stratosgear** [comment]
I think this really needs to be resolved.

I definitely do not want to share my workspace colors with any other user of the project, and I definitely would not want to "receive" a garish color from any of my colleagues (beauty is in the eye of the beholder, after all).

Peacock, is not really usable for me now, as is, because I cannot commit code when the workspace settings have been altered... :)

But I would really like to use it to differentiate different opened code instances...

How about that idea from [some posts back](https://github.com/johnpapa/vscode-peacock/issues/7#issuecomment-467220634)...?

**Comment by musicfuel** [comment]
@johnpapa I've been playing with this a bit focusing on the namespacing idea that Kenneth provided you with. Currently I've got it storing values in the User settings file so that it looks like:

```json
"peacock.workspaceColorCustomizations": {
    "/Users/musicfuel/Projects/vscode-ext/vscode-peacock/testworkspace": {
        "activityBar.background": "#1accff",
        "activityBar.foreground": "#15202b",
        "activityBar.inactiveForeground": "#15202b99",
        "activityBarBadge.background": "#df00ad",
        "activityBarBadge.foreground": "#e7e7e7",
        "titleBar.activeBackground": "#00b3e6",
        "titleBar.inactiveBackground": "#00b3e699",
        "titleBar.activeForeground": "#15202b",
        "titleBar.inactiveForeground": "#15202b99",
        "statusBar.background": "#00b3e6",
        "statusBarItem.hoverBackground": "#008bb3",
        "statusBar.foreground": "#15202b"
    }
},
```

The idea is that you could store a workspace's settings in this manner and then use activate/deactivate of the extension on application start, close, or workspace change to apply or clean up the customizations. The issue I can't get around is the whole concept of writing to `workbench.colorCustomizations` to override the colors. Doing so modifies the file and if the developer or team is not ignoring the file then it will certainly get inadvertently committed at some point.

Any ideas on how to apply Peacock coloring without modifying that file? I know in 1.32 some new support for applying themes came online, but I am pretty sure that is not available to extensions.

**Comment by johnpapa** [comment]
I'm not sure. 

Writing to user settings feels like it would be a big change ... meaning a lot to test and probably more than the tests currently cover. I'm open to it, but I sense some hurdles, such as what you point out.

Another idea is to create a `.peacockrc` file (as also suggested in some other issues by community folks). This doesn't solve it on its own, but at least then the changes are contained in one place. Then they would have to be applied somehow (maybe write them to the workspace settings and then clean them up?). It feels a bit error prone to me.

The only way to do this is to write to the settings file (that I know of). So whether it stores in user settings or another file, somehow they need to be applied. And that will update the workspace.

adding @auchenberg as he may have some ideas

**Comment by musicfuel** [comment]
Yes I agree. I don't like writing to the user settings either as that is a big change, but an aspect I do like of it is that it is scoped to the user and only the user. One issue with the `.peacockrc` approach is that we then still have to deal with how you have changes only impact you and not your other teammates when they pull the latest changes. I suppose that file could simply be part of the .gitignore (or other SCM ignore) and just be something that the team has to do to make this work.

In any of the cases, we are still left with the fragility of applying the settings and cleaning them up using the `workbench.colorCustomizations` approach. The code that looks for and applies those settings is in [colorThemeData.ts](https://github.com/Microsoft/vscode/blob/49c10b916705b6006f04d0701875098b26d1d70b/src/vs/workbench/services/themes/browser/colorThemeData.ts#L89) and it currently allows for two forms--the one we've been using and a slightly different syntax to override only a specific theme:

```json
"workbench.colorCustomizations": {
    "[Monokai]": {
        "statusBar.background": "#000"
    },
    "[Solarized]": {
        "statusBar.background": "#db57ac"
    }
}
```

This seems similar to what @auchenberg was referring to as an approach, but this is just for themes and doesn't really solve the problem of different workspaces and/or different users.

**Comment by johnpapa** [comment]
For reference, here is an issue in the VS Code repo that proposes a local workspace setting https://github.com/Microsoft/vscode/issues/40233#issuecomment-466390248

Quoting the original idea:

> So there would be three locations for settings (for a single-folder workspace):
> user settings
> - ${workspaceRoot}/.vscode/settings.json
> - ${workspaceRoot}/.vscode/settings.local.json

This would work well for our scenario, too. But I do not know if it is on the roadmap.

**Comment by musicfuel** [comment]
It's been open for quite some time. It does sound like something Peacock could very much take advantage of.

**Comment by auchenberg** [comment]
We don't have a good way to do this from a programatic context. You could leverage multi-root workspace settings, but that wouldn't be great for single workspaces. @johnpapa

**Comment by granmoe** [comment]
There was talk (in a mentioned VS Code issue, I believe) of having a third kind of vs code config that would be workspace-specific yet not shared. The objection to this was that it would complicate the settings story (understandable), but I disagree. It would be as simple as defining a .vscode-user/settings.json and then putting this in your .gitignore (or equivalent in other SCMs). Then you get workspace-specific user settings that are not shared, solving our issue.

That said, I think writing out the settings to normal user settings with color settings namespaced under each directory path is the ideal solution if possible.

**Comment by dkpalmer** [comment]
Just wanted to +1 the desire to manage in my own user settings and not in the workspace. I already know the rest of my team could never agree on a color 😅. Right now I'm just not checking in my peacock workspace settings.

**Comment by johnpapa** [comment]
> The idea is that you could store a workspace's settings in this manner and then use activate/deactivate of the extension on application start, close, or workspace change to apply or clean up the customizations. The issue I can't get around is the whole concept of writing to `workbench.colorCustomizations` to override the colors. Doing so modifies the file and if the developer or team is not ignoring the file then it will certainly get inadvertently committed at some point.
> 
> Any ideas on how to apply Peacock coloring without modifying that file? I know in 1.32 some new support for applying themes came online, but I am pretty sure that is not available to extensions.

@auchenberg is there a way to do this ^^^ ? To sum up, if we store the color settings in a namespaced property in user settings, we are thinking we still have to write them to workspace settings while the session is active, and wipe them when deactive. How can that be done? 

Or is there a way to have the colors be reda from the namespaced user settings?

**Comment by tommck** [comment]
I find it weird that vs code doesn't have a ".user" settings file like VS does. 

But, I would much prefer to have some file **outside** the workspace folder controlling this, since deleting a folder/re-retrieving from source control wouldn't wipe out my settings.

**Comment by TomDobbelaere** [comment]
+1 I don't want to use Peacock so long as it modifies the workspace settings.json :\ I don't want to accidentally check in the Peacock color changes, we do use our workspace settings file for other important settings so it's not like I can gitignore it

**Comment by mtwilliams5** [comment]
My team have inadvertendly come across this issue without even intending to use Peacock - Peacock gets installed as part of the Live Share extension pack, which we've recently started using, and as a result we keep getting `.vscode/settings.json` files created in the workspaces we're in. Half the time, these settings files are empty, or other times they seem to pull in colour customisation settings from the user settings.json file that were set outside of peacock for the user who happens to be hosting the live share at the time and aren't intended to be shared.
Since our working pattern normally ends up with us hopping between three or four different workspaces/repos within a piece of work, we don't want to go through and add `.vscode` to the gitignore for all of them, especially when this behaviour is being triggered by something that we're not even actively using...

**Comment by johnpapa** [comment]
@mtwilliams5 thanks for the feedback. I've added @lostintangent (from Live Share) and @egamma (from VS Code) to let them know some of the feedback you and others have reported on this.

There is not much more I can do at this time as Peacock writes to that file. If we had a way to write to a different file for workspace & user specific file that was auto-git-ignored, that would solve both issues mentioned here:
1. not wanting to git ignore the settings
2. not wanting to share the same colors

i'm open to any new ideas on how to solve this (see previous threads for what has been tried).

You can disable the peacock extension, too. 

One idea I hadn't thought of before ... because Peacock does write to this file when it detects it is newly installed. I could add a setting that effectively tells Peacock to be disabled and not write to the file. This feels wrong to me ... since there is a way to disable it already though ... so try that first.

**Comment by musicfuel** [comment]
I still think the ideal scenario would be for VS Code itself to allow for overriding the current theme colors without writing to the file. The need to write to a physical file is the issue that is tripping people up because it then leaves it on their teams to manage/ignore it or makes the logic of Peacock more complex with adding or removing it in a session.

If we could apply Peacock colors without the need to write to the file then it would be easy to change the extension to:

1. Always apply the style overrides to the current session virtually without the file
1. Add a command to write the current style to the workspace
   * Via user settings (possibly using the namespace approach mentioned above)
   * Via the workspace settings file so that it can be committed to the project for all of the team

I know this is not possible now, but I wonder if it could be made possible through a contribution to the VS Code code base.

**Comment by johnpapa** [comment]
That, @musicfuel , is a good idea! I know we discussed this before, but have you put this in the attached VS Code threads? i wonder if the team would be open to this. @egamma - would this be a possibility for a community member to help extend VS Code to allow it to change colors without writing to a file?

**Comment by musicfuel** [comment]
I don't think I ever wrote to the attached threads, but I know I looked into the code a bit at the time to see if there was any potential to add support for it. The https://github.com/microsoft/vscode/issues/40233#issuecomment-466390248 mentioned above has bits and pieces of it in the comments now, but doesn't seem like right issue to attach this to. I'll take a look at the existing issues around theme colors and see if there are any existing requests for this.

**Comment by octogonz** [comment]
@johnpapa Have you considered having Peacock write its color configuration into a `MyProject.code-workspace` file instead?  That file is a pretty good place for local user settings, and then the shared settings can go in `.vscode/settings.json`.  ([See here](https://code.visualstudio.com/docs/editor/multi-root-workspaces#_workspace-file) for docs.)

This has the slight disadvantage that you have to remember to open  the workspace file (`code MyProject.code-workspace`) instead of opening the folder (`code .`).  But VS Code reminds you about that when it sees a `.code-workspace` file.

The workspace approach also works for other user-specific settings such as **launch.json**.  And you can even have multiple `.code-workspace` files for the same repo folder.

**Comment by egamma** [comment]
@octogonz it is correct that you can put settings into the `code-workspace` file. However, the `code-workspace` file is typically checked into the repository and shared with the team. It therefore affects other users as well.

**Comment by jooeycheng** [comment]
Hi @johnpapa, would [VS Code Project Manager](https://github.com/alefragnani/vscode-project-manager#edit-projects)'s approach of having a local configuration file under the user settings work? It's essentially a central json keeping track of colors for each workspace.

**Comment by johnpapa** [comment]
Yes. Thanks for sharing that bill look at it. 

Storing project workspace settings in user settings can work but it has some downsides. I would have to copy them to workspace when you open a project and remove later. That still writes to workspace settings and can be buggy. 

Some other issues listed above too. 

But there is hope - local settings look like they are coming to vs code! The issue we linked to here is being worked on in April. I have no idea yet what form it will take or if or how it will help peacock. But - I am hopeful. 

So stay tuned.

**Comment by jooeycheng** [comment]
I see. Thanks for the prompt response!

> I would have to copy them to workspace when you open a project and remove later.

Ah, I didn't know that. So, we cannot escape writing to workspace config, given how VS Code works right now?

**Comment by johnpapa** [comment]
Vs code reads the setting to determine the colors. So it has to be in user settings or workspace settings. 

No worries

**Comment by gurgeous** [comment]
Great extension and I really want to use it! Just chiming in to say that I can't at this time because our team shares `.vscode/settings.json` and I can't commit to that file. That would be a gaffe on my part.

**Comment by mediantiba** [comment]
@johnpapa I love this extension but I'm hesitant to use it due to this very issue. I wouldn't want to force anyone to use my colors, let alone this extension if they didn't want to.

What if you write the config to some other file, e.g. `.vscode/peacock.json` which can be added to the git ignore list, wouldn't that solve the problem?

The extension could then read config based on priority, 1 being the highest and 3 being the lowest.
1. peacock settings (peacock.json)
2. workspace settings (settings.json)
3. user settings (global settings.json)

I don't know if I missed anything, but that way it would make it very easy to use the extension both individually and/or in a team environment :)

**Comment by musicfuel** [comment]
@mediantiba That is a good idea. Unfortunately the central problem behind this is not storing the settings in a manner that keeps them local and is instead about applying them to the interface of VSCode. Currently, the extension works by overriding the `workbench.colorCustomizations` setting in `.vscode/settings.json`. We could load settings for only you from the `peacock.json` file which is ignored, but it would then need to apply those changes to `.vscode/settings.json` which really shouldn't be ignored and typically is not. This makes a modification to that file which then can very easily be committed to the repo by accident.

Currently this mechanism is the only way for an extension to apply colors to the interface short of changing the theme completely which is not something this extension aims to do. Many solutions have been discussed, but all roads lead to changing `.vscode/settings.json` in the current session to actually apply the colors. There does appear to be ongoing work in the VSCode API to enable what Peacock wants to do to address this problem, but that hasn't landed into a release as of yet.

**Comment by johnpapa** [comment]
Well Said @musicfuel

**Comment by musicfuel** [comment]
To extend my comment above, see https://github.com/microsoft/vscode/issues/40233 for active discussion on how to deal with local settings. The simplest solution that many are arguing for (including @johnpapa) is to allow for some kind of settings.local.json file that could be created in the workspace folder. This file would then override workspace settings in the "default -> user -> workspace" settings inheritance. The file would allow for teams to still have common workspace settings, but individual users to override them. It would then be a determination of the user or team on how to manage that file in source control, but generally the idea is that someone savvy enough to create the file is also likely to setup the ignore support.

For some reason there is pushback to this approach in the thread, but I tend not to agree with the arguments and counter-proposal to it.

**Comment by mediantiba** [comment]
Thanks for the detailed explanation, @musicfuel, I wasn't aware of the limitations regarding this matter. Well, what can I say, I guess I hope to see local settings come to VS Code in a future update.

**Comment by prateek-redefcorp** [comment]
Bump!

**Comment by prateekkathal** [comment]
Bump!

**Comment by tom-sherman** [comment]
@prateekkathal please don't do that. There are a lot of people subscribed to this issue so that we can receive important updates, this type of comment is just spam in people's inboxes. See https://blog.sindresorhus.com/issue-bumping-e3b9740e2a0

**Comment by ivangretsky** [comment]
>  There does appear to be ongoing work in the VS Code API to enable what Peacock wants to do to address this problem, but that hasn't landed into a release as of yet.

@musicfuel , what are those features that are being worked on? We could at least go thumb them up)

**Comment by ThaDaVos** [comment]
I am currently running into this myself - something like a `settings.local.json` would solve it I think - as we want to commit our `settings.json` as it contains configurations for shared used extensions, like `i18n-ally` and stuff

**Comment by aisbergde** [comment]
I would prefer a separate file, which can be added into .gitignore
Then I will not disturb other users with my settings.
For example, I am using a bookmarks extension. It has its own file "bookmars.json" in the ".vscode" folder, and I can easily add it to .gitignore
Why not just use a file "peacock.json", which contains only the peacock settings?

**Comment by RexBarker** [comment]
So I'm here in 2023 with the same issue.  I just installed **peacock** and it looks great...functions how I would like it.  However, it's modifying my `settings.json` which is under git control (project wide).  From this loooong history here, I still don't understand if the issue was resolved.  Is there some other way to keep the peacock settings without using the `settings.json` file?

**Comment by carlocardella** [comment]
If all you care about is _user settings_ (`settings.json`) you could use a local only workspace (not under source control). For some of my repos I have a local workspace file (the file is saved in a separate folder, not under source control) that contains only the project folder. Color changes would be saved in this workspace file and will not affect other users unless you explicitly share the workpace file with them.

If you _do_ care about workspace settings (or yours are under source control) then you are in the same boat as everyone else here 😃.

I guess one could also experiment with the new [Profiles](https://code.visualstudio.com/updates/v1_74#_preview-features) (still in preview) since that creates a separate local settings file, not under source control.

**Comment by aisbergde** [comment]
Why it should be so complicated, instead of just using a separate file which can be included into the .gitignore? This would be intuitive and transparent.

Another idea is to gitignore the whole settings file. I think, it is better to use the general settings file for general project settings, and there should be a user specific setting file. But this is currently not the case in VSC. It would resolve some other issues with other settings to.

The profile solution seams to be complicated, but maybe it could be a way. I have no idea, how. Should the extension support profiles? Or how to define, which settings for which project will be stored in a profile?

**Comment by ivangretsky** [comment]
> Why it should be so complicated, instead of just using a separate file which can be included into the .gitignore? This would be intuitive and transparent.

> That being said, I think the ideal solution would be for Peacock to deal with the problem directly.

100% agreed

**Comment by dferber90** [comment]
I use the [Project Manager](https://marketplace.visualstudio.com/items?itemName=alefragnani.project-manager) extension all the time. It has a `projects.json` file (under `~/Library/Application Support/Code/User/globalStorage/alefragnani.project-manager/projects.json`) with a structure like

```json
[
  {
    "name": "flags",
    "rootPath": "/Users/xxx/Sites/happykit/flags",
    "paths": [],
    "tags": [],
    "enabled": true
  },
  // .... and more projects ...
]
```

This defines the projects available in VS Code that I can then quickly switch between.

It would be really useful if Peacock could store its settings in the same file, as my goal would be to use a different theme for each project. Using the same file has the benefit that the settings file is outside of any workspace, so no changes to .gitignore would be necessary. I'm not sure how technically possible this is, but it would be my ideal setup. If it can't be the same file, then a separate file for the Peacock extension itself would be the next best choice.

**Comment by dandax** [comment]
I don't think this is a Peacock problem at all. It's a very simple extension that builds on top of existing VS Code behavior. I've been following this issue (https://github.com/microsoft/vscode/issues/40233) for this exact reason. We should be looking at local/personalized settings instead of individual configuration. Local workspace settings would solve many more issues for everyone.

**Comment by benoitemile** [comment]
It might not be a Peacock issue but still, vscode issue you linked was created 5 years ago, we might need a solution until vscode solves it for good, there are plenty of possibilities listed here so we could avoid all those conflict, git stashing, git shash popping that very settings.json.

Like a .peacockrc that we could globally gitignore to avoid doing it in each repo.

**Comment by dandax** [comment]
> It might not be a Peacock issue but still, vscode issue you linked was created 5 years ago, we might need a solution until vscode solves it for good, there are plenty of possibilities listed here so we could avoid all those conflict, git stashing, git shash popping that very settings.json.
> 
> Like a .peacockrc that we could globally gitignore to avoid doing it in each repo.

Linking to a previous comment for reference https://github.com/johnpapa/vscode-peacock/issues/7#issuecomment-632707185. Also note the two following comments. That was May 2020.

**Comment by benoitemile** [comment]
Ok but I hope you got the point, it's 2.5years, maybe we could have a temporary solution until vscode provides desired behavior ?

**Comment by ivangretsky** [comment]
> > It might not be a Peacock issue but still, vscode issue you linked was created 5 years ago, we might need a solution until vscode solves it for good, there are plenty of possibilities listed here so we could avoid all those conflict, git stashing, git shash popping that very settings.json.
> > Like a .peacockrc that we could globally gitignore to avoid doing it in each repo.
> 
> Linking to a previous comment for reference [#7 (comment)](https://github.com/johnpapa/vscode-peacock/issues/7#issuecomment-632707185). Also note the two following comments. That was May 2020.

ok, but maybe 2.5 years later there is a hook that could be made to modify the settings read from `.settings` file with the ones read from `.peacockrc` before they get applied? Just a guess)

**Comment by nakah** [comment]
With the 1.75 release of VS Code, you can now have different profiles  (1 profile = 1 user settings.json).
To make it work, I've created one profile per workspace. Then you copy the settings generated by Peacock to your profile's settings.json.
I know it's not automated and it's more a workaround than a solution but it works !
You have your own color theme for your workspace without annoying your whole team 😊

**Comment by eckdanny** [comment]
Agree @nakah the [_recently-released_](https://code.visualstudio.com/updates/v1_75#_profiles) Profiles feature does not solve this. The Local Workspaces feature https://github.com/microsoft/vscode/issues/40233 (still open at the time of writing this) would be better. There exists an extension [Workspace Config+](https://marketplace.visualstudio.com/items?itemName=swellaby.workspace-config-plus) that seems to provide that _Local Workspace_ behavior (curious if anyone in this thread has tried yet?)

My workaround so far has been to add `.vscode/settings.json` to `<repository-root>/.git/info/exclude` after a clone (since it rarely changes if its tracked at all).

@johnpapa If Config+ is a more elegant workaround, could `Peacock` be configured to write to `settings.local.json` instead? Presumably, that'd be needed if `#40233` ever ships. I don't think the solution space here is in Peacock itself.

**Comment by acelinkio** [comment]
My primary use case is to differentiate between my open vscode windows.  Is there a way to just have Peacock set the current windows color randomly to and not write any configurations?  

I am happy with a low tech solutions.  If a color is reused, I can just close the window and reopen it again. I would rather not deal with managing state.

**Comment by NickBolles** [comment]
@johnpapa until the referenced local settings.json features hit in vscode, what do you think about adding a command to ignore the .vscode/settings.json file for convenience. Right now I'm running this command to ignore the settings.json changes since it's rarely changed (no idea if there are downsides to this yet).
```
git update-index --assume-unchanged ./.vscode/settings.json
```

A `Peacock: Ignore .vscode/settings.json changes` and `Peacock: UnIgnore .vscode/settings.json changes` (which would run with `--no-assume-unchanged` instead)

**Comment by fordneild** [comment]
https://marketplace.visualstudio.com/items?itemName=swellaby.workspace-config-plus

This extension solved the problem for me, by allowing me a settings.local.json

**Comment by jadnhm** [comment]
This issue has been open a long time and it seems not everyone is up to the task of reading the entire thread (as I just did...) so I thought I would grab this excellent summary of the crux of the problem by quoting @musicfuel 

https://github.com/johnpapa/vscode-peacock/issues/7#issuecomment-632707185
> ...Unfortunately the central problem behind this is not storing the settings in a manner that keeps them local and is instead about applying them to the interface of VS Code. Currently, the extension works by overriding the `workbench.colorCustomizations` setting in `.vscode/settings.json`. We could load settings for only you from the `peacock.json` file which is ignored, but it would then need to apply those changes to `.vscode/settings.json` which really shouldn't be ignored and typically is not. This makes a modification to that file which then can very easily be committed to the repo by accident.
> 
> Currently this mechanism is the only way for an extension to apply colors to the interface short of changing the theme completely which is not something this extension aims to do. Many solutions have been discussed, but all roads lead to changing `.vscode/settings.json` in the current session to actually apply the colors. There does appear to be ongoing work in the VS Code API to enable what Peacock wants to do to address this problem, but that hasn't landed into a release as of yet.

To paraphrase: the ONLY WAY to change the colours in VSCode at the present time is to _write_ the `"workbench.colorCustomizations"` to one of the official VSCode settings files

All the comments suggesting that 'Peacock could simply write to its own file!' are completely missing the point. 

It seems like some of the more recent commenters are having luck with [Workspace Config+](https://marketplace.visualstudio.com/items?itemName=swellaby.workspace-config-plus) and I am going to give that a whirl also, but this is still not a great solution because you're ultimately copying the `"workbench.colorCustomizations"` out of one of the official VSCode settings files (where Peacock knows to write those settings out) and into the special 'local' file so it can be utilized by the Workspace Config+ extension. 

We're all still crossing our fingers that https://github.com/microsoft/vscode/issues/40233 gets some traction and something is done about this by Microsoft/VSCode core team.

**Comment by georgeshevtsov** [comment]
git update-index --assume-unchanged .vscode/settings.json probably will do the trick for those that don't plan updating your settings.json anytime soon

**Comment by jadnhm** [comment]
yes and I use `git update-index --skip-worktree  .vscode/settings.json` but then you have to remember /undo/ that when you need to change that file which does sometimes happen....

**Comment by alecmce** [comment]
FWIW, I got here from https://stackoverflow.com/questions/72618635/visual-studio-code-workspace-settings-json-file-keeps-reverting-when-i-open-the after too long manually reverting settings.json. Glad to have finally found the cause!

**Comment by superjose** [comment]
Yesterday, I did an unintended April Fools. 

I thought the colors only affected the ones who had peacock installed.

Oh boy, I was wrong.

You gotta see the faces of all the team members when they saw that flashy pink on their screen.

Priceless 👌

**Comment by AferraroMezrah** [comment]
> This issue has been open a long time and it seems not everyone is up to the task of reading the entire thread (as I just did...) so I thought I would grab this excellent summary of the crux of the problem by quoting @musicfuel
> 
> [#7 (comment)](https://github.com/johnpapa/vscode-peacock/issues/7#issuecomment-632707185)
> 
> > ...Unfortunately the central problem behind this is not storing the settings in a manner that keeps them local and is instead about applying them to the interface of VS Code. Currently, the extension works by overriding the `workbench.colorCustomizations` setting in `.vscode/settings.json`. We could load settings for only you from the `peacock.json` file which is ignored, but it would then need to apply those changes to `.vscode/settings.json` which really shouldn't be ignored and typically is not. This makes a modification to that file which then can very easily be committed to the repo by accident.
> > Currently this mechanism is the only way for an extension to apply colors to the interface short of changing the theme completely which is not something this extension aims to do. Many solutions have been discussed, but all roads lead to changing `.vscode/settings.json` in the current session to actually apply the colors. There does appear to be ongoing work in the VS Code API to enable what Peacock wants to do to address this problem, but that hasn't landed into a release as of yet.
> 
> To paraphrase: the ONLY WAY to change the colours in VSCode at the present time is to _write_ the `"workbench.colorCustomizations"` to one of the official VSCode settings files
> 
> All the comments suggesting that 'Peacock could simply write to its own file!' are completely missing the point.
> 
> It seems like some of the more recent commenters are having luck with [Workspace Config+](https://marketplace.visualstudio.com/items?itemName=swellaby.workspace-config-plus) and I am going to give that a whirl also, but this is still not a great solution because you're ultimately copying the `"workbench.colorCustomizations"` out of one of the official VSCode settings files (where Peacock knows to write those settings out) and into the special 'local' file so it can be utilized by the Workspace Config+ extension.
> 
> We're all still crossing our fingers that [microsoft/vscode#40233](https://github.com/microsoft/vscode/issues/40233) gets some traction and something is done about this by Microsoft/VSCode core team.

But now with Profiles, don't they write to the official VSCode settings, and they change when you switch profiles? I'm pretty sure this feature was asked about before Profiles existed. But now that they have been released I was wondering if they could be the solution?

**Comment by jylenhof** [comment]
Maybe a solution is to use something like :
```
code --user-data-dir .vscode.local .
```
This create a User settings in .vscode.local/User/settings.json on each on your workspace
You put your own peacock configuration

You add .vscode.local to .gitignore

**Comment by axelra82** [comment]
> If there is a way to use profiles to do this no one has discovered/written about it yet AFAICT.
> 
> I just tried it again myself and it sure looks to me like this is still true:
> 
> > To paraphrase: the ONLY WAY to change the colours in VSCode at the present time is to write the "workbench.colorCustomizations" to one of the official VSCode settings files

Not sure I understand, with regard to `... the "workbench.colorCustomizations" to one of the official VSCode settings files`?

If you create a new profile in VSCode and assign it to the folder where your repo is, then set the `Settings` content source to a custom file  you are using `official VSCode settings`. **NOTE** you have to click the `Open to the side` file icon next to settings source when you hover it and `Create file`. This will create a profile specific user settings file.

### Screenshots of the VSCode profile settings
![code-profile-1](https://github.com/user-attachments/assets/f0f72aa3-2ee2-40fc-bf22-da33c41f3a57)
![code-profile-2](https://github.com/user-attachments/assets/91fd2b4b-2315-426b-82fe-bf7850ce2a82)


This is working fine for me, the only issue now is that `peacock` insist on updating the `.vscode/settings.json` file, with the `workbench.colorCustomizations` from my profile settings, i.e. as far as the workspace should be concerned, my `user settings` (a.k.a profile).

Maybe I'm missing something here, but I have to give :thumbs up: to @nakah (in https://github.com/johnpapa/vscode-peacock/issues/7#issuecomment-1426022933) for pointing this out.

Again, the "only" issue now (apart from this being a more manual solution than most might wish for) from what I can tell is that peacock is forcefully re-writing the `.vscode/settings.json` file when there's already a workaround using the profile (a.k.a "global user settings").

**Comment by jadnhm** [comment]
This is an exciting update @axelra82 !

I see a couple of 'requirements' for this that I would ask you about with this solution:

1. When I change my Peacock colour settings they are saved in a way that doesn't cause the source-control-tracked file `<project dir>/.vscode/settings.json` to be changed (ie: no need for `git update-index --assume-unchanged`)

2. When someone else changes the source-control-tracked file `<project dir>/.vscode/settings.json` to be changed on the 'remote' it will automatically be picked up by vscode when I do a regular source control update (ie: `git pull` or similar)


I feel like what you've described solves the first issue and not the second, but I haven't tried this yet so maybe I've missed some aspect of how profiles works. 


If this does present a possible solution, I think you've suggested a change to the Peacock internals to convince it to write its settings to a profile, yes? Could you describe how you think that should work?

**Comment by axelra82** [comment]
> This is an exciting update @axelra82 !
> 
> I see a couple of 'requirements' for this that I would ask you about with this solution:
> 
> 1. When I change my Peacock colour settings they are saved in a way that doesn't cause the source-control-tracked file `<project dir>/.vscode/settings.json` to be changed (ie: no need for `git update-index --assume-unchanged`)
> 2. When someone else changes the source-control-tracked file `<project dir>/.vscode/settings.json` to be changed on the 'remote' it will automatically be picked up by vscode when I do a regular source control update (ie: `git pull` or similar)
> 
> I feel like what you've described solves the first issue and not the second, but I haven't tried this yet so maybe I've missed some aspect of how profiles works.
> 
> If this does present a possible solution, I think you've suggested a change to the Peacock internals to convince it to write its settings to a profile, yes? Could you describe how you think that should work?

Well, like I stated earlier, "this works" for me. I.e. The only issue I'm currently dealing with is that peacock keeps forcing itself on the workspace settings file, so from time to time I have to remember to ignore (in git) the changes it has made to that file.

**NOTE** Using `git update-index --assume-unchanged` is not recommended, since you will have a file that **is** being changed but that **won't** show up in changed files in git 🤣 That's can be somewhat frustrating.

So really, _if peacock can just add a setting to stop writing to the settings_ file it will read the profile settings and that solves everything. No need to think about workspace settings in git sync (push/pull) as there's nothing there in regard to peacock since the local profile handles the coloring of workspaces.

**Comment by jadnhm** [comment]
FWIW I don't think peacock's code says `write to file " .vscode/settings.json"` -> it's using the vscode API to write to the settings file. I don't know how complicated it is to try to tell it /which/ settings file to write to, or what those APIs know about profiles... seems worth an investigation though

**Comment by escamoteur** [comment]
The biggest problem is that the color will even be set if another team member has the plugin not installed. I got this morning a pretty annoyed message from a team member

**Comment by gbro3n** [comment]
I was really excited to find this extension - but the workspace settings modification was an issue me too. I've built an extension that works around the limitation by temporarily changing user settings `workbench.colorCustomizations`, and then reverting when it looses focus. While not focused, the extension creates a coloured side bar so the VS Code instance can be identified while using `Alt + Tab` to switch windows. There's a quick video demo here https://github.com/appsoftwareltd/vscode-kingfisher, and the extension is here https://marketplace.visualstudio.com/items?itemName=AppSoftwareLtd.vscode-kingfisher

**Comment by johnpapa** [comment]
Thank you all for the long-running input here.

**Status:** keeping open as the umbrella issue for user-scoped / per-instance behavior that avoids committing Peacock colors to workspace settings.

Several related requests (including #495) should be consolidated here so design and implementation are not fragmented.

**Suggested labels:** enhancement, help wanted, umbrella, config-model.

**Suggested path:** contributors should propose a concrete first increment (data model + migration + fallback behavior) rather than trying to solve every variant in one PR.

**Minimal PR contract:**
- Add new user-scope configuration in package.json + src/models/enums.ts + src/configuration/read-configuration.ts
- Ensure behavior integrates cleanly in startup/apply flow (src/extension.ts + apply/config update paths)
- Add tests in src/test/ for precedence (user vs workspace), migration safety, and no-regression defaults
- Update docs/guide/ with migration guidance and docs/changelog/README.md

**Comment by strazto** [comment]
Migrating my proposal from #495 since it recommends a concrete pattern

**Is your feature request related to a problem? Please describe.**
I really like peacock for theming workspaces so I can tell my VSCode instances apart, but I don't necessarily want to check peacock theming into workspace settings.

I'd like an option to configure peacock profiles at the user level, and configure which workspaces should use those.

**Describe the solution you'd like**

I think a good way to do this might resemble `remote.SSH.remotePlatform", ie:

```json
    "remote.SSH.remotePlatform": {
        "windows": "windows",
        "server": "linux",
        "work": "linux",
        "work-externeal": "linux"
    },
```

By the above config, the remote SSH extension configures the platform for each of its hosts. An equivalent in peacock might be something like:

```json
{
    "peacock.workspaces.profiles": {
        "workspace-1-path-or-name?": "theme1",
        "workspace-2-path-or-name": "theme2",
       // etc
    }
    "peacock.profiles" : {
      "theme1": {
        "activityBar.activeBackground": "#0057b3",
        "activityBar.activeBorder": "#ff409d",
        "activityBar.background": "#0057b3",
        "activityBar.foreground": "#e7e7e7",
        "activityBar.inactiveForeground": "#e7e7e799",
        "activityBarBadge.background": "#ff409d",
        "activityBarBadge.foreground": "#15202b",
        "sash.hoverBorder": "#0057b3",
        "statusBar.background": "#003e80",
        "statusBar.foreground": "#e7e7e7",
        "statusBarItem.hoverBackground": "#0057b3",
        "statusBarItem.remoteBackground": "#003e80",
        "statusBarItem.remoteForeground": "#e7e7e7",
        "titleBar.activeBackground": "#003e80",
        "titleBar.activeForeground": "#e7e7e7",
        "titleBar.inactiveBackground": "#003e8099",
        "titleBar.inactiveForeground": "#e7e7e799"
    },
}  
```
**Describe alternatives you've considered**

 - Possibly using something like `.peacockrc` could work, as that can be ignored / not checked in separately to settings

I'm not sure about the best way to specify a workspace, there are a few options for pattern matching. 

Maybe follow some precedence based on explictness, eg, 
1. look for `peacockrc`, 
2. look for workspace setting `peacock.name`
3. match on folder base name
4. idk

**Comment by johnpapa** [comment]
Reviewing this one differently than the others, given the history — 80 comments, 206 reactions, and a real, substantial attempt already made and closed (#712).

For anyone landing here fresh: #712 was a genuine, well-researched 2000+ line implementation. It correctly identified that this has two separate problems — (1) storing a per-workspace color mapping outside `.vscode/settings.json`, which is straightforward, and (2) actually *applying* that color to a specific workspace without writing into that workspace's own settings file, which is the hard part, because VS Code's `workbench.colorCustomizations` has no scope between "this one folder's `.vscode/settings.json`" and "everything." The only technique #712 found that actually solves (2) is injecting custom CSS into the installed VS Code app itself (prior art: `jetbrains-titlebar`). It works, but it means Peacock would be patching VS Code's own files — real risk of VS Code flagging the install as corrupted, and of silently breaking on any VS Code update that changes the injected files' structure. It was closed without merging, which reads as that risk being a deliberate no-go rather than a quality problem with the PR itself.

Given that, I don't think re-attempting this is responsible without a decision on whether CSS injection's risk is actually acceptable for this extension — that's a call for @johnpapa to make explicitly, not something to quietly re-implement. If CSS injection is off the table, the honest state of this issue is: **blocked on a real VS Code API gap**, not on effort. Worth an explicit "won't do without X" or "here's the actual constraint" pinned comment given how much traffic this issue gets, so the next 80 comments aren't the same ground again.

Not opening a PR here — flagging the decision that's actually needed instead. 🙏

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by strazto** [comment]
Speaking as the implementer of the closed PR #712, I do agree with everything in the claude bot's comment https://github.com/johnpapa/vscode-peacock/issues/7#issuecomment-5595231186 

> (requires a) decision (from the maintainer) on whether CSS injection's risk is actually acceptable for this extension

> If CSS injection is off the table, blocked on a real VS Code API gap, 

> Worth an explicit "won't do without X" or "here's the actual constraint" pinned comment

**Comment by johnpapa** [comment]
Status note: this is honestly blocked on a real VS Code API gap — there's no clean way for an extension to apply theme colors per-user/per-instance without writing to a shared settings file that VS Code reads. Waiting to see if VS Code adds native support for this before investing further here. 🙏
